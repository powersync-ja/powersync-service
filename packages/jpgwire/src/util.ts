import * as tls from 'tls';
import { DEFAULT_CERTS } from './certs.js';
import * as pgwire from './pgwire.js';
import { PgType } from './pgwire_types.js';
import * as datefns from 'date-fns';

// TODO this is duplicated, but maybe that is ok
export interface NormalizedConnectionConfig {
  id: string;
  tag: string;

  hostname: string;
  port: number;
  database: string;

  username: string;
  password: string;

  sslmode: 'verify-full' | 'verify-ca' | 'disable';
  cacert: string | undefined;

  client_certificate: string | undefined;
  client_private_key: string | undefined;
}

type Mutable<T> = {
  -readonly [P in keyof T]: T[P];
};

export interface PgWireConnectionOptions extends NormalizedConnectionConfig {
  resolved_ip?: string;
}

export function clientTlsOptions(options: PgWireConnectionOptions): tls.ConnectionOptions {
  if (options.client_certificate && options.client_private_key) {
    return {
      cert: options.client_certificate,
      key: options.client_private_key
    };
  } else {
    return {};
  }
}
export function tlsOptions(options: PgWireConnectionOptions): false | tls.ConnectionOptions {
  if (options.sslmode == 'disable') {
    return false;
  } else if (options.sslmode == 'verify-full') {
    // Validate against default CA, or supplied CA.
    // Also validate hostname.
    return {
      ca: options.cacert ?? DEFAULT_CERTS,
      // hostname for TLS validation.
      // This may be different from the host we're connecting to if we pre-resolved
      // the IP.
      host: options.hostname,
      servername: options.hostname,
      ...clientTlsOptions(options)
    };
  } else if (options.sslmode == 'verify-ca') {
    if (options.cacert == null) {
      throw new Error(`cacert required for verify-ca`);
    }
    return {
      ca: options.cacert!,
      host: options.hostname,
      servername: options.hostname,
      // Disable hostname check
      checkServerIdentity: () => undefined,
      ...clientTlsOptions(options)
    };
  } else {
    throw new Error(`Unsupported sslmode: ${options.sslmode}`);
  }
}

export async function connectPgWire(config: PgWireConnectionOptions, options?: { type?: 'standard' | 'replication' }) {
  let connectionOptions: Mutable<ExtendedPgwireOptions> = {
    application_name: 'PowerSync',

    // tlsOptions below contains the original hostname
    hostname: config.resolved_ip ?? config.hostname,
    port: config.port,
    database: config.database,

    user: config.username,
    password: config.password
  };

  if (options?.type != 'standard') {
    connectionOptions.replication = 'database';
  }

  if (config.sslmode != 'disable') {
    connectionOptions.sslmode = 'require';

    // HACK: Not standard pgwire options
    // Just the easiest way to pass on our config to pgwire_node.js
    connectionOptions.sslrootcert = tlsOptions(config);
  } else {
    connectionOptions.sslmode = 'disable';
  }
  const connection = await pgwire.pgconnect(connectionOptions as pgwire.PgConnectOptions);
  // HACK: Replace row decoding with our own implementation
  (connection as any)._recvDataRow = _recvDataRow;
  return connection;
}

function _recvDataRow(this: any, _message: any, row: Uint8Array[], batch: any) {
  for (let i = 0; i < this._rowColumns.length; i++) {
    const valbuf = row[i];
    if (valbuf == null) {
      continue;
    }
    const { binary, typeOid } = this._rowColumns[i];
    // TODO avoid this._clientTextDecoder.decode for bytea
    row[i] = binary
      ? // do not valbuf.slice() because nodejs Buffer .slice does not copy
        // TODO but we not going to receive Buffer here ?
        Uint8Array.prototype.slice.call(valbuf)
      : PgType.decode(this._rowTextDecoder.decode(valbuf), typeOid);
  }
  batch.rows.push(row);
}

export interface PgPoolOptions {
  /**
   * Maximum number of open connections.
   *
   * Once this limit is reached, queries are queued.
   *
   * Defaults to 5.
   */
  maxSize?: number;

  /**
   * Idle timeout in ms before a connection is closed.
   */
  idleTimeout?: number | undefined;
}

/**
 * Open a postgres pool.
 *
 * The pool cannot be used for transactions, and cannot be used for logical replication.
 */
export function connectPgWirePool(config: PgWireConnectionOptions, options?: PgPoolOptions) {
  const idleTimeout = options?.idleTimeout;
  const maxSize = options?.maxSize ?? 5;

  let connectionOptions: Mutable<ExtendedPgwireOptions> = {
    application_name: 'PowerSync',

    // tlsOptions below contains the original hostname
    hostname: config.resolved_ip ?? config.hostname,
    port: config.port,
    database: config.database,

    user: config.username,
    password: config.password,

    _poolSize: maxSize,
    _poolIdleTimeout: idleTimeout
  };

  if (config.sslmode != 'disable') {
    connectionOptions.sslmode = 'require';

    // HACK: Not standard pgwire options
    // Just the easiest way to pass on our config to pgwire_node.js
    connectionOptions.sslrootcert = tlsOptions(config);
  } else {
    connectionOptions.sslmode = 'disable';
  }

  const pool = pgwire.pgpool(connectionOptions as pgwire.PgConnectOptions);
  const originalGetConnection = (pool as any)._getConnection;
  (pool as any)._getConnection = function (this: any) {
    const con = originalGetConnection.call(this);
    // HACK: Replace row decoding with our own implementation
    (con as any)._recvDataRow = _recvDataRow;
    return con;
  };
  return pool;
}

/**
 * Hack: sslrootcert is passed through as-is to pgwire_node.
 *
 * We use that to pass in custom TLS options, without having to modify pgwire itself.
 */
export interface ExtendedPgwireOptions extends pgwire.PgConnectKnownOptions {
  sslrootcert?: false | tls.ConnectionOptions;
}

export function lsnMakeComparable(text: string) {
  const [h, l] = text.split('/');
  return h.padStart(8, '0') + '/' + l.padStart(8, '0');
}

/**
 * Convert a postgres timestamptz to a SQLite-compatible/normalized timestamp.
 *
 * We output in the format:
 * YYYY-MM-DD HH:MM:SSZ
 * YYYY-MM-DD HH:MM:SS.SSSZ
 *
 * Precision is preserved from the source string, with the exception of ".000".
 *
 * We have specific exceptions for -infinity and infinity.
 */
export function timestamptzToSqlite(source?: string) {
  if (source == null) {
    return null;
  }
  // Make compatible with SQLite
  const match = /^([\d\-]+) ([\d:]+)(\.\d+)?([+-][\d:]+)$/.exec(source);
  if (match == null) {
    if (source == 'infinity') {
      return '9999-12-31 23:59:59Z';
    } else if (source == '-infinity') {
      return '0000-01-01 00:00:00Z';
    } else {
      return null;
    }
  }
  const [_, date, time, precision, timezone] = match as any;
  // datefns.parseISO() supports some dates that new Date() doesn't, e.g. '0022-09-08 21:19:28+00'
  const parsed = datefns.parseISO(`${date}T${time}${timezone}`);
  // Some dates are outside the range of JS and/or SQLite, e.g. 10022-09-08 21:19:28+00.
  // Just fallback to null for now.
  // We may consider preserving the source value in these cases.
  if (isNaN(parsed.getTime())) {
    return null;
  }
  const text = parsed.toISOString().replace('T', ' ').replace('.000', '').replace('Z', '');

  return `${text}${precision ?? ''}Z`;
}

/**
 * For timestamp without timezone, we keep it mostly as-is.
 *
 * We make specific exceptions for 'infinity' and '-infinity'.
 *
 * https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-DATETIME-SPECIAL-VALUES
 */
export function timestampToSqlite(source?: string) {
  if (source == null) {
    return null;
  }
  if (source == 'infinity') {
    return '9999-12-31 23:59:59';
  } else if (source == '-infinity') {
    return '0000-01-01 00:00:00';
  } else {
    return source;
  }
}
/**
 * For date, we keep it mostly as-is.
 *
 * We make specific exceptions for 'infinity' and '-infinity'.
 *
 * https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-DATETIME-SPECIAL-VALUES
 */
export function dateToSqlite(source?: string) {
  if (source == null) {
    return null;
  }
  if (source == 'infinity') {
    return '9999-12-31';
  } else if (source == '-infinity') {
    return '0000-01-01';
  } else {
    return source;
  }
}

/**
 * PgResult rows are arrays of values.
 *
 * This converts it to objects.
 */
export function pgwireRows(rs: pgwire.PgResult): Record<string, any>[] {
  const columns = rs.columns;
  return rs.rows.map((row) => {
    let r: Record<string, any> = {};
    for (let i = 0; i < columns.length; i++) {
      const c = columns[i];
      r[c.name] = row[i];
    }
    return r;
  });
}
