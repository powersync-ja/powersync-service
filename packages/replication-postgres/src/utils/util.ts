import * as pgwire from '@powersync/service-jpgwire';
import { replication } from '@powersync/service-core';

import { ReplicaIdentityResult } from '../types.js';
import { retriedQuery } from './pgwire_utils.js';

export function getReplicaIdColumns(relation: pgwire.PgoutputRelation): replication.ReplicationColumn[] {
  if (relation.replicaIdentity == 'nothing') {
    return [];
  } else {
    return relation.columns.filter((c) => (c.flags & 0b1) != 0).map((c) => ({ name: c.name, typeOid: c.typeOid }));
  }
}

export function getRelId(source: pgwire.PgoutputRelation): number {
  // Source types are wrong here
  const relId = (source as any).relationOid as number;
  if (relId == null || typeof relId != 'number') {
    throw new Error(`No relation id!`);
  }
  return relId;
}

export function getPgOutputRelation(source: pgwire.PgoutputRelation): replication.Relation {
  return {
    name: source.name,
    schema: source.schema,
    relationId: getRelId(source),
    replicaIdentity: source.replicaIdentity,
    replicationColumns: getReplicaIdColumns(source)
  };
}

export async function getPrimaryKeyColumns(
  db: pgwire.PgClient,
  relationId: number,
  mode: 'primary' | 'replident'
): Promise<replication.ReplicationColumn[]> {
  const indexFlag = mode == 'primary' ? `i.indisprimary` : `i.indisreplident`;
  const attrRows = await retriedQuery(db, {
    statement: `SELECT a.attname as name, a.atttypid as typeid, a.attnum as attnum
                                    FROM pg_index i
                                             JOIN pg_attribute a
                                                  ON a.attrelid = i.indrelid
                                                      AND a.attnum = ANY (i.indkey)
                                    WHERE i.indrelid = $1::oid
                                      AND ${indexFlag}
                                      AND a.attnum > 0
                                      ORDER BY a.attnum`,
    params: [{ value: relationId, type: 'int4' }]
  });

  return attrRows.rows.map((row) => {
    return { name: row[0] as string, typeOid: row[1] as number };
  });
}

export async function getAllColumns(db: pgwire.PgClient, relationId: number): Promise<replication.ReplicationColumn[]> {
  const attrRows = await retriedQuery(db, {
    statement: `SELECT a.attname as name, a.atttypid as typeid, a.attnum as attnum
                                    FROM pg_attribute a
                                    WHERE a.attrelid = $1::oid
                                      AND attnum > 0
                                    ORDER BY a.attnum`,
    params: [{ type: 'varchar', value: relationId }]
  });
  return attrRows.rows.map((row) => {
    return { name: row[0] as string, typeOid: row[1] as number };
  });
}

export async function getReplicationIdentityColumns(
  db: pgwire.PgClient,
  relationId: number
): Promise<ReplicaIdentityResult> {
  const rows = await retriedQuery(db, {
    statement: `SELECT CASE relreplident
        WHEN 'd' THEN 'default'
        WHEN 'n' THEN 'nothing'
        WHEN 'f' THEN 'full'
        WHEN 'i' THEN 'index'
     END AS replica_identity
FROM pg_class
WHERE oid = $1::oid LIMIT 1`,
    params: [{ type: 'int8', value: relationId }]
  });
  const idType: string = rows.rows[0]?.[0];
  if (idType == 'nothing' || idType == null) {
    return { replicationIdentity: 'nothing', columns: [] };
  } else if (idType == 'full') {
    return { replicationIdentity: 'full', columns: await getAllColumns(db, relationId) };
  } else if (idType == 'default') {
    return { replicationIdentity: 'default', columns: await getPrimaryKeyColumns(db, relationId, 'primary') };
  } else if (idType == 'index') {
    return { replicationIdentity: 'index', columns: await getPrimaryKeyColumns(db, relationId, 'replident') };
  } else {
    return { replicationIdentity: 'nothing', columns: [] };
  }
}
