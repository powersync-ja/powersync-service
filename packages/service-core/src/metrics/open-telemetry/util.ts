import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-http';
import { PrometheusExporter } from '@opentelemetry/exporter-prometheus';
import { MeterProvider, MetricReader, PeriodicExportingMetricReader } from '@opentelemetry/sdk-metrics';
import { logger } from '@powersync/lib-services-framework';
import { ServiceContext } from '../../system/ServiceContext.js';
import { MetricsFactory } from '../metrics-interfaces.js';
import { OpenTelemetryMetricsFactory } from './OpenTelemetryMetricsFactory.js';

import pkg from '../../../package.json' with { type: 'json' };
import { resourceFromAttributes } from '@opentelemetry/resources';

export function createOpenTelemetryMetricsFactory(context: ServiceContext): MetricsFactory {
  const { configuration, lifeCycleEngine, storageEngine } = context;
  const configuredExporters: MetricReader[] = [];

  if (configuration.telemetry.prometheus_port) {
    const prometheusExporter = new PrometheusExporter({
      port: configuration.telemetry.prometheus_port,
      preventServerStart: true
    });
    configuredExporters.push(prometheusExporter);

    lifeCycleEngine.withLifecycle(prometheusExporter, {
      start: async () => {
        await prometheusExporter.startServer();
        logger.info(`Prometheus metric export enabled on port:${configuration.telemetry.prometheus_port}`);
      }
    });
  }

  if (!configuration.telemetry.disable_telemetry_sharing) {
    const periodicExporter = new PeriodicExportingMetricReader({
      exporter: new OTLPMetricExporter({
        url: configuration.telemetry.internal_service_endpoint
      }),
      exportIntervalMillis: 1000 * 60 * 5 // 5 minutes
    });

    configuredExporters.push(periodicExporter);
  }

  let resolvedInstanceId: (id: string) => void;
  const instanceIdPromise = new Promise<string>((resolve) => {
    resolvedInstanceId = resolve;
  });

  lifeCycleEngine.withLifecycle(null, {
    start: async () => {
      const bucketStorage = storageEngine.activeBucketStorage;
      try {
        const instanceId = await bucketStorage.getPowerSyncInstanceId();
        resolvedInstanceId(instanceId);
      } catch (err) {
        resolvedInstanceId('Unknown');
      }
    }
  });

  const resource = resourceFromAttributes({
    ['service']: 'PowerSync',
    ['service.version']: pkg.version,
    ['instance_id']: instanceIdPromise
  });

  // This triggers OpenTelemetry to resolve the async attributes (instanceIdPromise).
  // This will never reject, and we don't specifically need to wait for it.
  resource.waitForAsyncAttributes?.();

  const meterProvider = new MeterProvider({
    resource,

    readers: configuredExporters
  });

  lifeCycleEngine.withLifecycle(meterProvider, {
    stop: async () => {
      await meterProvider.shutdown();
    }
  });

  const meter = meterProvider.getMeter('powersync');

  return new OpenTelemetryMetricsFactory(meter);
}
