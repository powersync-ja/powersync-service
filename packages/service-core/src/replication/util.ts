import * as pgwire from '@powersync/service-jpgwire';

import * as util from '../util/util-index.js';
import { ReplicationColumn, ReplicationIdentity } from './PgRelation.js';

export interface ReplicaIdentityResult {
  columns: ReplicationColumn[];
  replicationIdentity: ReplicationIdentity;
}

export async function getPrimaryKeyColumns(
  db: pgwire.PgClient,
  relationId: number,
  mode: 'primary' | 'replident'
): Promise<ReplicationColumn[]> {
  const indexFlag = mode == 'primary' ? `i.indisprimary` : `i.indisreplident`;
  const attrRows = await util.retriedQuery(db, {
    statement: `SELECT a.attname as name, a.atttypid as typeid, a.attnum as attnum
                                    FROM pg_index i
                                             JOIN pg_attribute a
                                                  ON a.attrelid = i.indrelid
                                                      AND a.attnum = ANY (i.indkey)
                                    WHERE i.indrelid = $1::oid
                                      AND ${indexFlag}
                                      AND a.attnum > 0
                                      ORDER BY a.attnum`,
    params: [{ value: relationId, type: 'int4' }]
  });

  return attrRows.rows.map((row) => {
    return { name: row[0] as string, typeOid: row[1] as number };
  });
}

export async function getAllColumns(db: pgwire.PgClient, relationId: number): Promise<ReplicationColumn[]> {
  const attrRows = await util.retriedQuery(db, {
    statement: `SELECT a.attname as name, a.atttypid as typeid, a.attnum as attnum
                                    FROM pg_attribute a
                                    WHERE a.attrelid = $1::oid
                                      AND attnum > 0
                                    ORDER BY a.attnum`,
    params: [{ type: 'varchar', value: relationId }]
  });
  return attrRows.rows.map((row) => {
    return { name: row[0] as string, typeOid: row[1] as number };
  });
}

export async function getReplicationIdentityColumns(
  db: pgwire.PgClient,
  relationId: number
): Promise<ReplicaIdentityResult> {
  const rows = await util.retriedQuery(db, {
    statement: `SELECT CASE relreplident
        WHEN 'd' THEN 'default'
        WHEN 'n' THEN 'nothing'
        WHEN 'f' THEN 'full'
        WHEN 'i' THEN 'index'
     END AS replica_identity
FROM pg_class
WHERE oid = $1::oid LIMIT 1`,
    params: [{ type: 'int8', value: relationId }]
  });
  const idType: string = rows.rows[0]?.[0];
  if (idType == 'nothing' || idType == null) {
    return { replicationIdentity: 'nothing', columns: [] };
  } else if (idType == 'full') {
    return { replicationIdentity: 'full', columns: await getAllColumns(db, relationId) };
  } else if (idType == 'default') {
    return { replicationIdentity: 'default', columns: await getPrimaryKeyColumns(db, relationId, 'primary') };
  } else if (idType == 'index') {
    return { replicationIdentity: 'index', columns: await getPrimaryKeyColumns(db, relationId, 'replident') };
  } else {
    return { replicationIdentity: 'nothing', columns: [] };
  }
}
