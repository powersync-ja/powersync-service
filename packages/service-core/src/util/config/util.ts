import * as urijs from 'uri-js';

export function getSupabaseJwksUrl(connection: any) {
  if (connection == null) {
    return null;
  } else if (connection.type != 'postgresql') {
    return null;
  }

  let hostname: string | undefined = connection.hostname;
  if (hostname == null && typeof connection.uri == 'string') {
    hostname = urijs.parse(connection.uri).host;
  }
  if (hostname == null) {
    return null;
  }

  const match = /db.(\w+).supabase.co/.exec(hostname);
  if (match == null) {
    return null;
  }
  const projectId = match[1];

  return `https://${projectId}.supabase.co/auth/v1/.well-known/jwks.json`;
}
