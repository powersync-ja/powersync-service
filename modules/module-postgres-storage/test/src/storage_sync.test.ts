import { register } from '@powersync/service-core-tests';
import { describe } from 'vitest';
import { POSTGRES_STORAGE_FACTORY } from './util.js';

/**
 * Bucket compacting is not yet implemented.
 * This causes the internal compacting test to fail.
 * Other tests have been verified manually.
 */
describe('sync - postgres', () => {
  register.registerSyncTests(POSTGRES_STORAGE_FACTORY);
});
