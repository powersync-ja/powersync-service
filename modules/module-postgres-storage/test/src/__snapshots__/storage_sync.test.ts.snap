// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sync - postgres > compacting data - invalidate checkpoint 1`] = `
[
  {
    "checkpoint": {
      "buckets": [
        {
          "bucket": "mybucket[]",
          "checksum": -93886621,
          "count": 2,
          "priority": 3,
        },
      ],
      "last_op_id": "2",
      "write_checkpoint": undefined,
    },
  },
]
`;

exports[`sync - postgres > compacting data - invalidate checkpoint 2`] = `
[
  {
    "data": {
      "after": "0",
      "bucket": "mybucket[]",
      "data": [
        {
          "checksum": -93886621n,
          "op": "CLEAR",
          "op_id": "2",
        },
      ],
      "has_more": false,
      "next_after": "2",
    },
  },
  {
    "checkpoint_diff": {
      "last_op_id": "4",
      "removed_buckets": [],
      "updated_buckets": [
        {
          "bucket": "mybucket[]",
          "checksum": 499012468,
          "count": 4,
          "priority": 3,
        },
      ],
      "write_checkpoint": undefined,
    },
  },
  {
    "data": {
      "after": "2",
      "bucket": "mybucket[]",
      "data": [
        {
          "checksum": 1859363232n,
          "data": "{"id":"t1","description":"Test 1b"}",
          "object_id": "t1",
          "object_type": "test",
          "op": "PUT",
          "op_id": "3",
          "subkey": "02d285ac-4f96-5124-8fba-c6d1df992dd1",
        },
        {
          "checksum": 3028503153n,
          "data": "{"id":"t2","description":"Test 2b"}",
          "object_id": "t2",
          "object_type": "test",
          "op": "PUT",
          "op_id": "4",
          "subkey": "a17e6883-d5d2-599d-a805-d60528127dbd",
        },
      ],
      "has_more": false,
      "next_after": "4",
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "4",
    },
  },
]
`;

exports[`sync - postgres > expired token 1`] = `
[
  {
    "token_expires_in": 0,
  },
]
`;

exports[`sync - postgres > expiring token 1`] = `
[
  {
    "checkpoint": {
      "buckets": [
        {
          "bucket": "mybucket[]",
          "checksum": 0,
          "count": 0,
          "priority": 3,
        },
      ],
      "last_op_id": "0",
      "write_checkpoint": undefined,
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "0",
    },
  },
]
`;

exports[`sync - postgres > expiring token 2`] = `
[
  {
    "token_expires_in": 0,
  },
]
`;

exports[`sync - postgres > sends checkpoint complete line for empty checkpoint 1`] = `
[
  {
    "checkpoint": {
      "buckets": [
        {
          "bucket": "mybucket[]",
          "checksum": -1221282404,
          "count": 1,
          "priority": 3,
        },
      ],
      "last_op_id": "1",
      "write_checkpoint": undefined,
    },
  },
  {
    "data": {
      "after": "0",
      "bucket": "mybucket[]",
      "data": [
        {
          "checksum": 3073684892,
          "data": "{"id":"t1","description":"sync"}",
          "object_id": "t1",
          "object_type": "test",
          "op": "PUT",
          "op_id": "1",
          "subkey": "02d285ac-4f96-5124-8fba-c6d1df992dd1",
        },
      ],
      "has_more": false,
      "next_after": "1",
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "1",
    },
  },
  {
    "checkpoint_diff": {
      "last_op_id": "1",
      "removed_buckets": [],
      "updated_buckets": [],
      "write_checkpoint": "1",
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "1",
    },
  },
]
`;

exports[`sync - postgres > sync buckets in order 1`] = `
[
  {
    "checkpoint": {
      "buckets": [
        {
          "bucket": "b0[]",
          "checksum": 920318466,
          "count": 1,
          "priority": 2,
        },
        {
          "bucket": "b1[]",
          "checksum": -1382098757,
          "count": 1,
          "priority": 1,
        },
      ],
      "last_op_id": "2",
      "write_checkpoint": undefined,
    },
  },
  {
    "data": {
      "after": "0",
      "bucket": "b1[]",
      "data": [
        {
          "checksum": 2912868539n,
          "data": "{"id":"earlier","description":"Test 2"}",
          "object_id": "earlier",
          "object_type": "test",
          "op": "PUT",
          "op_id": "2",
          "subkey": "243b0e26-87b2-578a-993c-5ac5b6f7fd64",
        },
      ],
      "has_more": false,
      "next_after": "2",
    },
  },
  {
    "partial_checkpoint_complete": {
      "last_op_id": "2",
      "priority": 1,
    },
  },
  {
    "data": {
      "after": "0",
      "bucket": "b0[]",
      "data": [
        {
          "checksum": 920318466n,
          "data": "{"id":"t1","description":"Test 1"}",
          "object_id": "t1",
          "object_type": "test",
          "op": "PUT",
          "op_id": "1",
          "subkey": "02d285ac-4f96-5124-8fba-c6d1df992dd1",
        },
      ],
      "has_more": false,
      "next_after": "1",
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "2",
    },
  },
]
`;

exports[`sync - postgres > sync global data 1`] = `
[
  {
    "checkpoint": {
      "buckets": [
        {
          "bucket": "mybucket[]",
          "checksum": -93886621,
          "count": 2,
          "priority": 3,
        },
      ],
      "last_op_id": "2",
      "write_checkpoint": undefined,
    },
  },
  {
    "data": {
      "after": "0",
      "bucket": "mybucket[]",
      "data": [
        {
          "checksum": 920318466n,
          "data": "{"id":"t1","description":"Test 1"}",
          "object_id": "t1",
          "object_type": "test",
          "op": "PUT",
          "op_id": "1",
          "subkey": "02d285ac-4f96-5124-8fba-c6d1df992dd1",
        },
        {
          "checksum": 3280762209n,
          "data": "{"id":"t2","description":"Test 2"}",
          "object_id": "t2",
          "object_type": "test",
          "op": "PUT",
          "op_id": "2",
          "subkey": "a17e6883-d5d2-599d-a805-d60528127dbd",
        },
      ],
      "has_more": false,
      "next_after": "2",
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "2",
    },
  },
]
`;

exports[`sync - postgres > sync legacy non-raw data 1`] = `
[
  {
    "checkpoint": {
      "buckets": [
        {
          "bucket": "mybucket[]",
          "checksum": -852817836,
          "count": 1,
          "priority": 3,
        },
      ],
      "last_op_id": "1",
      "write_checkpoint": undefined,
    },
  },
  {
    "data": {
      "after": "0",
      "bucket": "mybucket[]",
      "data": [
        {
          "checksum": 3442149460n,
          "data": {
            "description": "Test
"string"",
            "id": "t1",
            "large_num": 12345678901234567890n,
          },
          "object_id": "t1",
          "object_type": "test",
          "op": "PUT",
          "op_id": "1",
          "subkey": "02d285ac-4f96-5124-8fba-c6d1df992dd1",
        },
      ],
      "has_more": false,
      "next_after": "1",
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "1",
    },
  },
]
`;

exports[`sync - postgres > sync updates to data query only 1`] = `
[
  {
    "checkpoint": {
      "buckets": [
        {
          "bucket": "by_user["user1"]",
          "checksum": 0,
          "count": 0,
          "priority": 3,
        },
      ],
      "last_op_id": "1",
      "write_checkpoint": undefined,
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "1",
    },
  },
]
`;

exports[`sync - postgres > sync updates to data query only 2`] = `
[
  {
    "checkpoint_diff": {
      "last_op_id": "2",
      "removed_buckets": [],
      "updated_buckets": [
        {
          "bucket": "by_user["user1"]",
          "checksum": 1418351250,
          "count": 1,
          "priority": 3,
        },
      ],
      "write_checkpoint": undefined,
    },
  },
  {
    "data": {
      "after": "0",
      "bucket": "by_user["user1"]",
      "data": [
        {
          "checksum": 1418351250n,
          "data": "{"id":"list1","user_id":"user1","name":"User 1"}",
          "object_id": "list1",
          "object_type": "lists",
          "op": "PUT",
          "op_id": "2",
          "subkey": "5ad0aa14-3d5e-5428-ad5b-2c33927d991c",
        },
      ],
      "has_more": false,
      "next_after": "2",
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "2",
    },
  },
]
`;

exports[`sync - postgres > sync updates to global data 1`] = `
[
  {
    "checkpoint": {
      "buckets": [
        {
          "bucket": "mybucket[]",
          "checksum": 0,
          "count": 0,
          "priority": 3,
        },
      ],
      "last_op_id": "0",
      "write_checkpoint": undefined,
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "0",
    },
  },
]
`;

exports[`sync - postgres > sync updates to global data 2`] = `
[
  {
    "checkpoint_diff": {
      "last_op_id": "1",
      "removed_buckets": [],
      "updated_buckets": [
        {
          "bucket": "mybucket[]",
          "checksum": 920318466,
          "count": 1,
          "priority": 3,
        },
      ],
      "write_checkpoint": undefined,
    },
  },
  {
    "data": {
      "after": "0",
      "bucket": "mybucket[]",
      "data": [
        {
          "checksum": 920318466n,
          "data": "{"id":"t1","description":"Test 1"}",
          "object_id": "t1",
          "object_type": "test",
          "op": "PUT",
          "op_id": "1",
          "subkey": "02d285ac-4f96-5124-8fba-c6d1df992dd1",
        },
      ],
      "has_more": false,
      "next_after": "1",
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "1",
    },
  },
]
`;

exports[`sync - postgres > sync updates to global data 3`] = `
[
  {
    "checkpoint_diff": {
      "last_op_id": "2",
      "removed_buckets": [],
      "updated_buckets": [
        {
          "bucket": "mybucket[]",
          "checksum": -93886621,
          "count": 2,
          "priority": 3,
        },
      ],
      "write_checkpoint": undefined,
    },
  },
  {
    "data": {
      "after": "1",
      "bucket": "mybucket[]",
      "data": [
        {
          "checksum": 3280762209n,
          "data": "{"id":"t2","description":"Test 2"}",
          "object_id": "t2",
          "object_type": "test",
          "op": "PUT",
          "op_id": "2",
          "subkey": "a17e6883-d5d2-599d-a805-d60528127dbd",
        },
      ],
      "has_more": false,
      "next_after": "2",
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "2",
    },
  },
]
`;

exports[`sync - postgres > sync updates to parameter query + data 1`] = `
[
  {
    "checkpoint": {
      "buckets": [],
      "last_op_id": "0",
      "write_checkpoint": undefined,
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "0",
    },
  },
]
`;

exports[`sync - postgres > sync updates to parameter query + data 2`] = `
[
  {
    "checkpoint_diff": {
      "last_op_id": "2",
      "removed_buckets": [],
      "updated_buckets": [
        {
          "bucket": "by_user["user1"]",
          "checksum": 1418351250,
          "count": 1,
          "priority": 3,
        },
      ],
      "write_checkpoint": undefined,
    },
  },
  {
    "data": {
      "after": "0",
      "bucket": "by_user["user1"]",
      "data": [
        {
          "checksum": 1418351250n,
          "data": "{"id":"list1","user_id":"user1","name":"User 1"}",
          "object_id": "list1",
          "object_type": "lists",
          "op": "PUT",
          "op_id": "1",
          "subkey": "5ad0aa14-3d5e-5428-ad5b-2c33927d991c",
        },
      ],
      "has_more": false,
      "next_after": "1",
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "2",
    },
  },
]
`;

exports[`sync - postgres > sync updates to parameter query only 1`] = `
[
  {
    "checkpoint": {
      "buckets": [],
      "last_op_id": "0",
      "write_checkpoint": undefined,
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "0",
    },
  },
]
`;

exports[`sync - postgres > sync updates to parameter query only 2`] = `
[
  {
    "checkpoint_diff": {
      "last_op_id": "1",
      "removed_buckets": [],
      "updated_buckets": [
        {
          "bucket": "by_user["user1"]",
          "checksum": 0,
          "count": 0,
          "priority": 3,
        },
      ],
      "write_checkpoint": undefined,
    },
  },
  {
    "checkpoint_complete": {
      "last_op_id": "1",
    },
  },
]
`;
